[ConstraintSystem] Diagnose conditional requirement failures via fixes

xedin opened the pull request:
Extend existing `RequirementFailure` functionality to support
conditional requirement failures. Such fixes are introduced
only if the parent type requirement has been matched successfully.

Resolves: rdar://problem/47871590

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Add requirement kind to conditional conformance requirement locator

xedin committed:
[CSDiagnostics] Augment RequirementFailure to support conditional requirements

xedin committed:
[ConstraintSystem] Diagnose conditional requirement failures via fixes

Extend existing `RequirementFailure` functionality to support
conditional requirement failures. Such fixes are introduced
only if the parent type requirement has been matched successfully.

Resolves: rdar://problem/47871590

xedin committed:
[Sema] NFC: Add accessors to check whether locator is type or conditional requirement

xedin commented on test/Generics/conditional_conformances.swift:
```diff
@@ -425,3 +391,14 @@ protocol P {
+    return self <= 1
+            ? 1
+            : (2...self).reduce(1, *)
+            // expected-error@-1 {{instance method 'reduce' requires that 'Self.Stride' conform to 'SignedInteger'}}
```
This diagnostic has a note:

```
note: requirement from conditional conformance of 'ClosedRange<Self>' to 'Sequence'
extension ClosedRange : Sequence where Bound : Strideable, Bound.Stride : SignedInteger {
          ^
```

slavapestov commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -88,19 +90,54 @@ Type RequirementFailure::getOwnerType() const {
+  // reference.
+  if (isConditional()) {
+    auto *conformance = getConformanceRef().getConcrete();
+    auto *DC = conformance->getDeclContext();
```
We should think of a better way to express this. I think I've see this appear a few times in this file.

xedin replied:
Actually I realized that I can simply do `getRequirementDC()` in this case, no need to fetch conformance explicitly, thanks!

xedin replied:
Hm, after thinking about this - it genericContext() should be related to the affect decl here and not the requirement, so this if shouldn't even be needed...

xedin replied:
Fixed!

xedin replied:
That also improved diagnostic for:
```swift
extension BinaryInteger {
  var factorial: Self {
    return self <= 1
                ? 1
                : (2...self).reduce(1, *)
  }
}
```

To:
```
error: referencing instance method 'reduce' on 'ClosedRange' requires that 'Self.Stride' conform to 'SignedInteger'
                  : (2...self).reduce(1, *)
                               ^
note: requirement from conditional conformance of 'ClosedRange<Self>' to 'Sequence'
extension ClosedRange : Sequence where Bound : Strideable, Bound.Stride : SignedInteger {
          ^
```

slavapestov commented on test/Constraints/array_literal.swift:
```diff
@@ -338,7 +339,7 @@ extension Int: P { }
 func testConditional(i: Int, s: String) {
   let _: PArray<Int> = [i, i, i]
-  let _: PArray<String> = [s, s, s] // expected-error{{cannot convert value of type '[String]' to specified type 'PArray<String>'}}
+  let _: PArray<String> = [s, s, s] // expected-error{{generic struct 'PArray' requires that 'String' conform to 'P'}}
```
Cool!

xedin committed:
[CSDiagnostics] Improve requirement failure source detection

If affected declaration is a static or instance memeber (excluding
operators) and failed requirement is declared in other nominal type
or extension, diagnose such problems as `in reference` instead of
claiming that requirement belongs to the member itself.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@shahmishal Looks like this is something with the bot itself in the final phase.

xedin commented:
@swift-ci please test

harlanhaskins commented on test/Generics/conditional_conformances.swift:
```diff
@@ -12,55 +12,36 @@ protocol P6: P2 {}
+// expected-note@-1 {{candidate requires that the types 'U' and 'V' be equivalent (requirement specified as 'U' == 'V')}}
+// expected-note@-2 {{requirement from conditional conformance of 'SameTypeGeneric<U, V>' to 'P2'}}
+// expected-note@-3 {{requirement from conditional conformance of 'SubclassBad' to 'P2'}}
+// expected-note@-4 {{candidate requires that 'C1' inherit from 'U' (requirement specified as 'U' : 'C1')}}
```
Should this read `candidate requires that 'U' inherit from 'C1'`?

xedin replied:
Yeah, probably, I didn't touch existing messages in declaration checker.
